README.rst
=======
EEA PDF
=======
.. image:: https://ci.eionet.europa.eu/buildStatus/icon?job=eea/eea.pdf/develop
  :target: https://ci.eionet.europa.eu/job/eea/job/eea.pdf/job/develop/display/redirect
  :alt: develop
.. image:: https://ci.eionet.europa.eu/buildStatus/icon?job=eea/eea.pdf/master
  :target: https://ci.eionet.europa.eu/job/eea/job/eea.pdf/job/master/display/redirect
  :alt: master

This package allows your users to download Plone content as PDF files. As admin
you can define specific PDF themes per content-type.

.. note ::

  Requires `wkhtmltopdf`_ system-package installed on your server.
  See `eea.converter`_ documentation for more details.


Contents
========

.. contents::


Main features
=============

1. Adds download as PDF action at the bottom of the page
2. Possibility to define custom PDF themes per content-type
3. Asynchronously generate PDF files and notify users by email when PDF is ready
4. Possibility to temporarily disable dynamic PDF creation by adding an item
   called 'action-download-pdf' within context


Install
=======

- Add eea.pdf to your **eggs** and **zcml** section in your buildout
  and re-run buildout.
  You can download a sample buildout from
  https://github.com/eea/eea.pdf/tree/master/buildouts/plone4
- Install eea.pdf within Site Setup > Add-ons

::

    [instance]
    eggs =
        ...
        eea.pdf

    zcml =
        ...
        eea.pdf


External PDF generator tools
----------------------------
Ensure that you have installed `wkhtmltopdf`_ on your machine. You
can also install `wkhtmltopdf`_ from buildout::

    [buildout]

    parts +=
        wkhtmltopdf

    [wkhtmltopdf]
    recipe = hexagonit.recipe.download
    url = http://eggrepo.apps.eea.europa.eu/pypi/wkhtmltopdf/wkhtmltopdf-0.12.1.tgz

    [instance]
    environment-vars +=
        WKHTMLTOPDF_PATH ${wkhtmltopdf:location}/wkhtmltopdf

Asynchronous setup
------------------
By default all PDFs are generated asynchronous, therefore some extra config is
needed within your buildout in order for this to work properly.

First of all you'll need a folder were to store generated PDF files. For this
you can create it manually within buildout:directory/var/ or you can let buildout
handle it::

    [buildout]

    parts +=
        media-downloads
        media-downloads-temp


    media-downloads-path = ${buildout:directory}/var/downloads/pdf
    media-downloads-temp = ${buildout:directory}/var/downloads/tmp

    [media-downloads]
    recipe = ore.recipe.fs:mkdir
    path = ${buildout:media-downloads-path}
    mode = 0700
    createpath = true

    [media-downloads-temp]
    recipe = ore.recipe.fs:mkdir
    path = ${buildout:media-downloads-temp}
    mode = 0700
    createpath = true

This will create a folder named **downloads** within buildout:directory/var/

Next, in order for this folder to be visible from your website and your users to
be able to download generated PDFs you'll need to tell to your zope instances
about it::

    [buildout]

    media-downloads-name = downloads
    media-downloads-path = ${buildout:directory}/var/downloads/pdf
    media-downloads-temp = ${buildout:directory}/var/downloads/tmp

    [instance]

    environment-vars +=
        EEADOWNLOADS_NAME ${buildout:media-downloads-name}
        EEADOWNLOADS_PATH ${buildout:media-downloads-path}
        EEACONVERTER_TEMP ${buildout:media-downloads-temp}

Also, don't forget to setup `plone.app.async`_

::

    [buildout]

    [instance]
    eggs +=
        plone.app.async
    zcml +=
        plone.app.async-single_db_worker

Getting started
===============

1. Go to Site Setup > PDF Settings
2. Customize an existing PDF theme or add a new one
3. Go to Home page and click on download as pdf icon at the bottom of the page
   or directly access http://localhost:8080/Plone/front-page/download.pdf


PDF Theme options
=================
For more information about wkhtmltopdf options like Table of contents XSL file,
header and footer parameters, etc. see `wkhtmltopdf`_ documentation.

Cover
-----
A page template to be used for PDF Cover. Leave empty to disable it.
Default: pdf.cover

Disclaimer
----------
A page template containing copyright and author information. This page will be
placed between cover and PDF body. Leave empty to skip it.
Default: pdf.disclaimer

Body
----
A page template to be used for PDF body. An empty value will skip PDF body.
Default: pdf.body

Back Cover
----------
A page template to be used for back cover. Leave empty to disable back cover.
Default: pdf.cover.back

Header
------
A page template to be used as PDF body header. This will not appear on cover,
disclaimer or back cover. Leave empty for no header.
Default: pdf.header

Footer
------
A page template to be used as PDF body footer. This will not appear on cover,
disclaimer or back.cover. Leave empty for no footer.
Default: pdf.footer

Table of contents
-----------------
An XSL page template to be used for PDF Table of contents. See `wkhtmltopdf`_
documentation for more information about XSL format. Leave empty to disable
Table of contents.
Default: pdf.toc

Table of contents links
-----------------------
Enable or disable Table of Contents internal links and also
PDF bookmarks (outline)
Default: False

JavaScript
----------
Enable or disable javascript.
Default: True

JavaScript Delay
----------------
Wait some seconds for javascript to finish
Default: 0

Timeout
-------
Abort PDF conversion after this number of seconds
Default: 3600

Asynchronous
------------
Generate PDF asynchronously and send an email to the user when it's done
Default: True

Offset
------
Start counting pages within PDF Body from this number. Usefull when cover and/or
disclaimer are enabled.
Default: 0

Maximum depth
-------------
This option defines the maximum depth a folderish item can go while recursively
includes it's children within PDF.
Default: 1 (include only direct children, non-folderish ones)

Maximum breadth
---------------
This options limit the number of direct children a folderish item can include
within PDF.
Default: 100

Maximum items
-------------
The total items to be included within PDF export for a folderish item, including
depth and breadth.
Default: 1000

Portal types
------------
Apply this theme to selected portal types.
Default:


Custom permissions
==================
Custom permissions added by this package

Can download PDF (eea.pdf.download)
-----------------------------------
Assign this permission to roles that you want to be able to download content as PDF
Default: Owner, Manager, Editor

Can customize PDF (eea.pdf.customize)
-------------------------------------
Assign this permission to roles that you want to be able to contextually customize
the output PDF look and feel
Default: Manager, Site Administrator

Content rules
=============
This package uses Plone Content-rules to notify users by email when an asynchronous
PDF job is done. Thus 3 custom content-rules will be added within
Plone > Site Setup > Content-rules

.. warning ::

  As these content-rules are triggered by an asynchronous job, while
  you customize the email template for these content-rules,
  please **DO NOT USE OTHER** string substitutions **that the ones** that start
  with **$download_** as you'll break the download chain.
  Also if you disable these content-rules the users will never know when the
  PDF is ready and what is the link where they can download the output PDF.

Export succeeded
----------------
Notify the person who requested a PDF export that the PDF successfully exported
and provide a link to the downloadable PDF

Export failed
-------------
Notify the person who requested a PDF export that the PDF export failed.

Export failed (admin)
---------------------
Notify admin that there were issues while exporting PDF


Content rules email string substitution
=======================================
In order to be able to easily customize emails sent by this package the following
custom email template string substitutions can be made


${download_came_from_url}
-------------------------
The absolute URL of the Plone object which is downloaded as PDF

${download_email}
-----------------
Email address of the user that triggered the download as PDF action

${download_error}
-----------------
Error traceback when download as PDF job fails

${download_from_email}
----------------------
Site Admin email address customizable via Plone > Site Setup > Mail

${download_from_name}
---------------------
Site Admin name customizable via Plone > Site Setup > Mail

${download_title}
-----------------
Title of the Plone object which is downloaded as PDF

${download_url}
---------------
The absolute URL where the generated output PDF can be downloaded

${download_type}
----------------
Download type. Default to PDF for this package. It is package specific and it
can be PDF, EPUB, etc.


Disable PDF export
==================
You have the possibility to temporarily disable dynamic PDF export contextually
by adding a static PDF file (or a Python Script, Page Template, etc)
within context called **action-download-pdf**. This way /download.pdf will
return this file instead of generating one based on context data.

.. note::

  This works only with folderish items.

Troubleshooting
===============
PDFs are generated asynchronously using a parallel zc.async queue.quota.
The number of workers that will generate PDFs in parallel is automatically
calculated based on the number of zeo-clients registered with
**plone.app.async-*_db_worker**.

As every **db_worker** can handle simultaneously **maximum 3 jobs** (hard-coded in zc.async Agent),
if you have **2 workers** then the maximum number of PDFs that will be generated
at the same time will be **6** (2 workers * 3). Same if you have **5**, you'll get
**15 PDFs** generated at the same time.

If for any reason you don't want them to be generated simultaneously you can set
environment variable **EEAPDF_ASYNC_THREADS** to **1** within buildout::

    [buildout]

    ...

    [instance]

    ...

    environment-vars +=
        EEAPDF_ASYNC_THREADS 1


Also, if you experience issues by having too many simultaneously PDF jobs, you
can limit them in the same way as above.


Dependencies
============

1. `eea.converter`_
2. `eea.downloads`_
3. `wkhtmltopdf`_
4. `plone.app.async`_
5. `eea.cache`_ (optional)

Source code
===========

- Latest source code (Plone 4 compatible):
  https://github.com/collective/eea.pdf


Copyright and license
=====================
The Initial Owner of the Original Code is European Environment Agency (EEA).
All Rights Reserved.

The EEA PDF (the Original Code) is free software;
you can redistribute it and/or modify it under the terms of the GNU
General Public License as published by the Free Software Foundation;
either version 2 of the License, or (at your option) any later
version.

More details under docs/License.txt


Funding
=======

EEA_ - European Environment Agency (EU)

.. _EEA: https://www.eea.europa.eu/
.. _eea.converter: https://eea.github.io/docs/eea.converter
.. _eea.downloads: https://eea.github.io/docs/eea.downloads
.. _wkhtmltopdf: http://wkhtmltopdf.org
.. _eea.cache: https://eea.github.io/docs/eea.cache
.. _plone.app.async: https://github.com/plone/plone.app.async#ploneappasync


docs/HISTORY.txt
Changelog
=========

9.6-dev0 - (unreleased)
---------------------------

9.5 - (2021-07-14)
---------------------------
* Bugfix: un-hide entire content-header within print
  [ichim-david refs #136489]

9.4 - (2021-05-11)
---------------------------
* Bugfix: use filename from the pdf uploaded in pdf static field otherwise
  default to download.pdf
  [ichim-david refs #133890]

9.3 - (2020-09-18)
--------------------------
* Bugfix: Adjust pdf output of indicators
  [szabozo0 refs #121188]

9.2 - (2019-07-04)
--------------------------
* Change: Use sans serif fallback instead of Open Sans due to issues with font loading of wkhtmltopdf 
  [ichim-david refs #107145]

9.1 - (2019-01-28)
--------------------------
* Jenkins: Add sonarqube step
  [avoinea refs #101552]

9.0 - (2018-11-23)
--------------------------
* Change: Group only the first paragraph with heading
  h2 and h3 within "nobreak" container to avoid to many
  white-space blocks in generated PDF files.
  [avoinea refs #100179]
* Bug fix: Fixed PDF header width for Fiches
  [avoinea refs #100179]

8.9 - (2018-08-30)
--------------------------
* Bug fix: Fixed PDF generator with short title
  [avoinea refs #97618]


8.8 - (2018-08-21)
--------------------------
* Change: updated URLs pointing to eea.europa.eu with https://
  [alecghica refs #95849]

8.7 - (2018-05-25)
-----------------------
* Bug fix: Speed-up cache invalidation for backward-references
  [avoinea refs #95020]

8.6 - (2018-04-24)
-----------------------
* Change: Make zc.async logs less verbose about PDF downloads
  [avoinea]

8.5 - (2018-01-08)
------------------
* Change: updated pdf.print.css with changes from custom
  [ichim-david refs #90180]

8.4 - (2017-12-12)
------------------
* Change: Replace eeacms/zptlint with eeacms/plone-test:4 zptlint
  [avoinea refs #90415]

8.3 - (2017-11-07)
------------------
* Change: Remove Sphinx generated documentation
  [petchesi-iulian refs #88212]

8.2 - (2017-07-10)
------------------
* Change: can_download now checks also zope2.View permission on context.
  This avoids a situation where you cannot view the context but you can
  trigger a pdf download
  [ichim-david refs #86581]

8.1 - (2017-06-23)
------------------
* Bug fix: return None if the attribute is not found
  [iulianpetchesi #86181]

8.0 - (2017-05-22)
------------------
* Change: fixed PyLint warnings and errors
  [valipod refs #84949]

7.9 - (2017-05-15)
------------------
* Change: fixed PyLint warnings and errors
  [eduard-fironda refs #84949]

7.8 - (2017-04-24)
------------------
* Change: updated package information
  [eduard-fironda]

7.7 - (2017-04-19)
------------------
* Bug fix: safely check for cover subtitle
  [ichim-david refs #83520]

7.6 - (2017-04-13)
------------------
* Feature: added ability to disable cover subtitle per theme
  [ichim-david refs #83520]

7.5 - (2017-03-02)
------------------
* Bugfix: fix wkhtmltopdf segmentation fault by setting 100% height
  on img which have page-break-inside avoid
  [ichim-david refs #80861]

7.4 - (2017-02-21)
------------------
* Change: Update pdf.print CSS and JS with more elements to be hidden
  [avoinea refs #81054]

7.3 - (2016-12-05)
------------------
* Feature: added translations
  [irina-botez]

7.2 - (2016-11-14)
------------------
* Change: word break external-link avoiding thus having a smaller
  pdf render
  [ichim-david refs #78802]

7.1 - (2016-11-07)
------------------
* Feature: added css classes for cover-header types
  [ichim-david refs #78802]
* Bug fix: fixed figures hidden when placed after featureHeading
  [ichim-david refs #78802]

7.0 - (2016-10-31)
------------------
* Bug fix: fixed pdf output of collection pdf theme where title
  and page had a lot of empty space
  [ichim-david refs #77970]
* Bug fix: fixed figure heading from being disconnected from the
  figure charts that follow them
  [ichim-david refs #77970]

6.9 - (2016-10-14)
------------------
* Feature: book theme now shows folder title within toc and content
  [ichim-david refs #77476]
* Bug fix: Remove image property from profile exporting
  [ZuluPro refs #76260]

6.8 - (2016-10-03)
------------------
* Change: Updated to use plone.stringinterp IContextWrapper adapter
  [avoinea]
* Bug fix: include children of the folders when generate pdf
  [catalin refs #75057]

6.7 - (2016-09-05)
------------------
* Change: PDFTheme now uses a blob image field
  [ichim-david refs #75045]

6.6 - (2016-02-15)
------------------
* Change: fixed pdf disclaimer qrbox margins placement
  [ichimdav refs #69095]

6.5 - (2015-12-08)
------------------
* Change: Use Open sans for figure titles and googlecharts-notes of
  inline charts that are part of indicators
  [ichimdav refs #30473]

6.4 - (2015-12-07)
------------------
* Change: content table is now of display table instead of block
  in order to strech tables to 100% of viewable space
  [ichimdav refs #30473]

6.3 - (2015-11-16)
------------------
* Change: box-classification is now visible in the pdf output
  [ichimdav refs #30421]

6.2 - (2015-10-22)
------------------
* Bug fix: pdf segmentation fault
  [lucas refs #28298]
* Feature: Possibility to bulk invalidate PDFs per theme
  [chiridra refs #28517]

6.1 - (2015-10-05)
------------------
* Bug fix: avoid a page break immediately after a header
  [lucas refs #28588]

6.0 - (2015-09-10)
------------------
* Change: Drop pdftk dependency and use pure Python PyPDF2 package
  [voineali refs #28522]

5.9 - (2015-09-04)
------------------
* Change: removed indicator-figure styling for pdf.print.css, any fixes to the
  indicator pdf sizes can be added in other eea packages
  [ichimdav refs #28278]

5.8 - (2015-08-18)
------------------
* Feature: Switched to dedicated zc.async queue quota for PDF jobs which allows
  multiple jobs to run in parallel. This will speed up PDF generation by using
  all plone.app.async worker instances at the same time.
  [voineali refs #28019]

eea/pdf/content/theme.py
""" EEA Relations Content Type
"""
from zope.interface import implementer
from Products.Archetypes import atapi
from Products.ATContentTypes.content.folder import ATFolder
from plone.app.blob.field import ImageField
from eea.pdf.content.interfaces import IPDFTheme
from eea.pdf.config import EEAMessageFactory as _
from archetypes.referencebrowserwidget import ReferenceBrowserWidget

EditSchema = atapi.Schema((
    atapi.StringField('cover',
        schemata='default',
        default='pdf.cover',
        widget=atapi.StringWidget(
            label=_(u"Cover"),
            description=_(
                u"A page template to be used for PDF Cover. "
                u"Leave empty to disable it."
            )
        )
    ),
    atapi.BooleanField('coverSubtitle',
        schemata='default',
        default=True,
        widget=atapi.BooleanWidget(
           label=_(u"Cover subtitle"),
           description=_(u"Enable display of subtitle within cover page."
                         u" By default subtitle is created from content"
                         u" description truncated to first 70 characters"
            )
       )
    ),

    atapi.StringField('disclaimer',
        schemata='default',
        default='pdf.disclaimer',
        widget=atapi.StringWidget(
            label=_(u"Disclaimer"),
            description=_(
                u"A page template to be used for PDF Disclaimer, "
                u"the first page after PDF cover. "
                u"Leave empty to disable it."
            )
        )
    ),
    atapi.StringField('body',
        schemata='default',
        default='pdf.body',
        widget=atapi.StringWidget(
            label=_(u"Body"),
            description=_(
                u"A page template to be used for PDF Body. "
                u"Leave empty to disable it."
            )
        )
    ),
    atapi.StringField('backcover',
        schemata='default',
        default='pdf.cover.back',
        widget=atapi.StringWidget(
            label=_(u"Back Cover"),
            description=_(
                u"A page template to be used for PDF Back Cover. "
                u"Leave empty to disable it."
            )
        )
    ),
    atapi.StringField('header',
        schemata='default',
        default='pdf.header',
        widget=atapi.StringWidget(
            label=_(u"Header"),
            description=_(
                u"A page template to be used for PDF Header. "
                u"Leave empty to disable it."
            )
        )
    ),
    atapi.StringField('footer',
        schemata='default',
        default='pdf.footer',
        widget=atapi.StringWidget(
            label=_(u"Footer"),
            description=_(
                u"A page template to be used for PDF Footer. "
                u"Leave empty to disable it."
            )
        )
    ),

    atapi.BooleanField('staticFooterAndHeader',
        schemata='default',
        default=False,
        widget=atapi.BooleanWidget(
           label=_(u"Static Footer and Header"),
           description=_(u"Enable static footer and header html pages. "
                         u"Footer and header templates are evaluated once"
                         u" and then served for every pdf page."
            )
       )
    ),

    atapi.StringField('toc',
        schemata='default',
        default='pdf.toc',
        widget=atapi.StringWidget(
            label=_(u"Table of contents"),
            description=_(
                u"An XSL page template to be used for PDF Table of contents. "
                u"Leave empty to disable it."
            )
        )
    ),

    atapi.BooleanField('toclinks',
       schemata='default',
       default=False,
       widget=atapi.BooleanWidget(
           label=_(u"Table of contents links"),
           description=_(u"Enable table of contents links")
       )
    ),
    atapi.ReferenceField(
        'imagescollection',
        relationship='imagescollectionrel',
        multiValued=0,
        allowed_types=('Collection', 'ATTopic', 'Folder'),
        widget=ReferenceBrowserWidget(
            allow_search=True,
            allow_browse=True,
            force_close_on_insert=True,
            label=_(u"Images cover collection"),
            description=_(u"if cover image not present use "
                           "a random cover from the collection."),
        ),
    ),
    atapi.BooleanField('javascript',
       schemata='default',
       default=True,
       widget=atapi.BooleanWidget(
           label=_(u"JavaScript"),
           description=_(u"Enable or disable javascript")
       )
    ),
    atapi.IntegerField('javascriptdelay',
       schemata='default',
       default=0,
       widget=atapi.IntegerWidget(
           label=_(u"JavaScript delay"),
           description=_(u"Wait some seconds for javascript to finish")
       )
    ),
    atapi.IntegerField('timeout',
        schemata='default',
        default=3600,
        widget=atapi.IntegerWidget(
            label=_(u"Timeout"),
            description=_(
                u"Abort PDF export after specified number of seconds. "
                u"Use zero to disable it."
            )
        )
    ),
    atapi.BooleanField('async',
        schemata='default',
        default=True,
        widget=atapi.BooleanWidget(
            label=_(u"Asynchronous"),
            description=_(
                u"Generate PDF asynchronously and send an email to the user "
                u"when it's done."
            )
        )
    ),
    atapi.IntegerField('offset',
        schemata='default',
        default=0,
        widget=atapi.IntegerWidget(
            label=_(u"Offset"),
            description=_(
                u"Page numbering offset within PDF Body."
            )
        )
    ),
    atapi.IntegerField('maxdepth',
        schemata='default',
        default=1,
        widget=atapi.IntegerWidget(
            label=_(u"Maximum depth"),
            description=_(
                u"Maximum depth to recursively include children items "
                u"while generating PDF for collection "
                u"or folderish content-types."
            )
        )
    ),
    atapi.IntegerField('maxbreadth',
        schemata='default',
        default=100,
        widget=atapi.IntegerWidget(
            label=_(u"Maximum breadth"),
            description=_(
                u"Maximum breadth to include children items "
                u"while generating PDF for collection "
                u"or folderish content-types where object is collection or "
                u"folderish."
            )
        )
    ),
    atapi.IntegerField('maxitems',
        schemata='default',
        default=1000,
        widget=atapi.IntegerWidget(
            label=_(u"Maximum items"),
            description=_(
                u"Total maximum children items to be included "
                u"while generating PDF for collection "
                u"or folderish content-types."
            )
        )
    ),
    atapi.LinesField('types',
         schemata='default',
         vocabulary_factory='plone.app.vocabularies.ReallyUserFriendlyTypes',
         multiValued=1,
         widget=atapi.MultiSelectionWidget(
             format='checkbox',
             label=_(u'Portal types'),
             description=_(
                 u"Use this PDF Theme for the following Portal-Types."
             )
         )
    ),
    ImageField("image",
        schemata="default",
        sizes=None,
        widget=atapi.ImageWidget(
            label=_(u"Preview"),
            description=_(u"Upload a preview image for this theme")
        )
    ),
))

THEME_SCHEMA = ATFolder.schema.copy() + EditSchema.copy()

def finalize_schema(schema=THEME_SCHEMA):
    """ Update schema
    """
    for field in schema.fields():
        field.write_permission = 'Manage portal'
        if field.schemata != 'default':
            field.required = False
            field.mode = 'r'

finalize_schema()

@implementer(IPDFTheme)
class PDFTheme(ATFolder):
    """ PDF Theme
    """
    portal_type = meta_type = 'PDFTheme'
    archetypes_name = 'EEA PDF Theme'
    _at_rename_after_creation = True
    schema = THEME_SCHEMA
    schema['description'].widget.modes = ()
